osgar/drivers/deedee.py

```diff
@@ -0,0 +1,165 @@
+            buf.write(bytes(written))
+        # Finish the message.
+        buf.write(bytes([SLIP_END]))
+        self.publish('raw', buf.getvalue())
+
+
+    def on_raw(self, data):
+        packets = self.slip_stream.send(data)
+        for packet in packets:
+            self.publish('packet', packet)
+
+
+class Deedee(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        bus.register('cmd', 'pose2d', 'emergency_stop', 'stdout')
+        self.emergency_stop_status = None
+        self.wheel_base = config.get('wheel_base', 0.298)
+        self.max_command_validity = config.get('max_command_validity', 0.8)
+        self.control_timeout = datetime.timedelta(seconds=config.get('control_timeout', 1.0))
+        self.last_cmd_time = None
+
+
+    def update(self):
+        timestamp, channel, data = self.bus.listen()
+        if channel == 'desired_speed':
+            self.on_speed_cmd(timestamp, data)
+        elif channel == 'info':
+            self.on_info(data)
+        elif channel == 'tick' and (
+                self.last_cmd_time is None or
+                timestamp - self.last_cmd_time >= self.control_timeout):
+            self.send_speed_cmd(timestamp, 0, 0)
+
+
+    def send_speed_cmd(self, timestamp, forward, rotation):
+        # Send the speed command.
+        delta = self.wheel_base * rotation / 2
+        left, right = round((forward - delta) * 1000), round((forward + delta) * 1000)
+        validity = round(self.max_command_validity * 1000)
+        cmd = struct.pack('>Bhhh', ord('V'), left, right, validity)
+        checksum = -sum(cmd) & 0xFF
+        cmd += bytes([checksum])
+        self.publish('cmd', cmd)
+        self.last_cmd_time = timestamp
+
+        # Immediately ask for the current state.
+        req = ord('?')
+        checksum = -req & 255
+        cmd = bytes([req, checksum])
+        self.publish('cmd', cmd)
+
+
+    def on_speed_cmd(self, timestamp, msg):
+        forward, rotation = msg
+        self.send_speed_cmd(timestamp, forward / 1000, math.radians(rotation / 100))
+
+
+    def on_info(self, msg):
+        checksum = sum(msg) & 0xFF
+        if checksum != 0:
+            print('Incorrect checksum: {}'.format(checksum))
+            return
+
+        if msg[0] == ord('!'):
+            (x, y, phi, emergency_stop) = struct.unpack('>qqq?', msg[1:-1])
+            self.publish('pose2d',
+                         [round(x * 1e-3), round(y * 1e-3),
+                             round(math.degrees(phi * 1e-4))])
+            emergency_stop = bool(emergency_stop)
+            if emergency_stop != self.emergency_stop_status:
+                self.publish('emergency_stop', emergency_stop)
+                self.emergency_stop_status = emergency_stop
+        elif msg[0] == ord('E') and msg[:-1] != b'E\nOK V\n':
+            print('Communication error: {}'.format(msg[:-1]))
+            self.publish('stdout', msg[:-1].decode('ascii'))
+
+class Demo(Node):
```

Comment:
the same Demo should work for `python -m osgar.go run config/deedee-demo.json` ... in the second interation ...